### PR TITLE
Added destroyColumn migration step

### PR DESCRIPTION
**Pull request by bmatasar**
I don't know how to add info about dropped columns in getSyncChanges

### CODE REVIEW

**Comment by radex** [comment]
Nice! Thank you for your contribution! Some quick thoughts:

- `getSyncChanges` provides summary of migration changes via sync to backend. The purpose of that is to allow server to update client's database with tables and columns it could not previously sync as client had no way of expressing them. There's no reason to notify backend about removed columns or tables, since there's nothing extra for server to send to client to align the two schemas.
- there's some formatting issues - try running `yarn prettier` on the codebase to fix

File: src/Schema/migrations/test.js [hunk 1]
```diff
@@ -181,6 +181,10 @@ describe('migration step functions', () => {
       'type',
     )
   })
+  it('throws if destroyColumn() is malformed', () => {
```
Comment (radex) [inline src/Schema/migrations/test.js:184]: pls update one of the tests above to test for the format of what destroyColumn returns

File: src/adapters/lokijs/worker/DatabaseDriver.js [hunk 2]
```diff
@@ -423,6 +426,15 @@ export default class DatabaseDriver {
     })
   }
 
+  _executeDestroyColumnMigration({ table, column }: DestroyColumnMigrationStep): void {
+    const collection = this.loki.getCollection(table)
+
+    // update ALL records in the collection, adding new fields
```
Comment (radex) [inline src/adapters/lokijs/worker/DatabaseDriver.js:432]:
```suggestion
    // update ALL records in the collection, remove a field
```

File: src/Schema/migrations/getSyncChanges/index.js [hunk 3]
```diff
@@ -45,13 +50,33 @@ export default function getSyncChanges(
...
+    .reduce((addedColumns, step) => {
+      if (!createdTables.includes(step.table)) {
+        switch (step.type) {
+          case 'add_columns':
+            return addedColumns.concat(
+              step.columns.map(({ name }) => ({ table: step.table, name })),
+            )
+          case 'destroy_column':
+            return addedColumns.filter(
+              ({ table, name }) => step.table !== table || step.column !== name,
```
Comment (radex) [inline src/Schema/migrations/getSyncChanges/index.js:72]: this doesn't seem correct to me, and there are no tests for this. can you explain?

**Review by radex — CHANGES_REQUESTED** [review]
Okay, I reviewed it, and have a few requests:

- Please update docs-master/Advanced/Migrations.md
- add an integration test to adapters/__tests__/commonTEsts.js - you'll notice that there are some migration tests there that actually test if new columns/tables are usable after the addition. You could modify them or add a new test that tests if destroyed columns really were destroyed

Otherwise, it all seems reasonable, thanks for your work!

File: src/Schema/migrations/getSyncChanges/index.js (hunk 3 above)
Comment (bmatasar) [reply src/Schema/migrations/getSyncChanges/index.js:72]: I had wrong understanding of getSyncChanges. I still have.
- If there is an addColumns for 'table' and 'column' in an earlier step, a destroyColumn in a later step should cancel it out. That's why I unnested them and kept them as a single array.

**Comment by bmatasar** [comment]
It looks like ALTER TABLE DROP COLUMN is not supported although the docs say it does. However, ALTER TABLE RENAME COLUMN is supported. I am stuck right now :)

**Comment by bmatasar** [comment]
I have just executed `yarn ci:check` and it fails on my mac.

**Comment by radex** [comment]
oh no https://sqlite.org/changes.html this was only added in 3.35.0 :(

**Comment by radex** [comment]
you could go with something like this : https://stackoverflow.com/questions/5938048/delete-column-from-sqlite-table but you probably have to drop and recreate indices for it to work

**Comment by bmatasar** [comment]
That's what I was afraid of. I would need to pass also the whole table description in order to do that. It makes the whole migration step really ugly.
What sqlite3 version is watermelondb supporting?

**Comment by radex** [comment]
@bmatasar Unfortunately, it's complicated. On iOS and Android we're using the system built-in version of sqlite, so we can't rely on newer sqlite features. On Android JSI we're bundling sqlite 3.36 https://github.com/Nozbe/WatermelonDB/blob/master/package.json#L48 because it's impossible to reuse system version of sqlite. We're also using `better-sqlite3` for Node support of SQLite (this is a rare feature probably used by just two people + 🍉's testing setup), I'm not sure which version of sqlite3 it's building, you need to check.

I would prefer not to bundle sqlite3 on iOS due to codesize/performance concerns

**Comment by bmatasar** [comment]
The problem is that in order to apply the solution with the duplication of the table I need the whole schema for the table. The statements are generated just from the migrations steps, so it will look more like a `keepColumns` step listing only the remaining columns.
I guess I need more time to think about it.

**Comment by radex** [comment]
yeah that's fair. I forgot about this, but this difficulty is why I never implemented dropping tables and columns in migrations - it shouldn't be very hard but it's not trivial

**Comment by bmatasar** [comment]
Let's close the PR for now.

**Comment by radex** [comment]
I'm gonna re-open so the work done already doesn't get lost

**Comment by radex** [comment]
@bmatasar read about the magic sqlite tables, I think you can get info about indices with pure SQL

**Comment by vonovak** [comment]
just passing by, hope you don't mind the notification
I wanted to say that this should be good to integrate soon. [ios 15](https://stackoverflow.com/questions/14288128/what-version-of-sqlite-does-ios-provide) ships sqlite version 3.36.0

now with ios 17 being out, I believe requiring ios 15 won't be an issue 👍

**Comment by bmatasar** [comment]
I would rather start a new PR based on #1654

**Pull request closed without merging**